### PR TITLE
Fix sync early exit bug. Track new session during sync. Simplify thre…

### PR DESCRIPTION
**Pull request by plorenz**
…ading and tighten up locking. Add dump api-sessions debug operation

### CODE REVIEW

File: controller/server/controller.go [hunk 1]
```diff
@@ -263,9 +263,9 @@ func (c *Controller) Run() {
 		pfxlog.Logger().Fatalf("failed to create Edge Client API factory: %v", err)
 	}
 
-	if err = c.policyEngine.Start(c.AppEnv.HostController.GetCloseNotifyChannel()); err != nil {
-		log.WithError(err).Fatalf("error starting policy engine")
-	}
+	//if err = c.policyEngine.Start(c.AppEnv.HostController.GetCloseNotifyChannel()); err != nil {
```
Comment (andrewpmartinez) [inline controller/server/controller.go:266]: Commented out?

File: controller/sync_strats/sync_instant.go [hunk 2]
```diff
@@ -446,6 +446,7 @@ func (strategy *InstantStrategy) synchronize(rtx *RouterSender) {
 
 			if len(apiSessions) >= chunkSize {
 				state.IsLast = !cursor.IsValid()
+				logger.Infof("sending api session sync chunk %v, isLast=%v", state.Sequence, state.IsLast)
```
Comment (andrewpmartinez) [inline controller/sync_strats/sync_instant.go:449]: Were these meant to be `Info`?

File: controller/server/controller.go (hunk 1 above)
Comment (plorenz) [reply controller/server/controller.go:266]: For testing purposes, need to keep the massive set of expired api tokens around. Will put back in before merging. Please don't approve until it's back in ;)

File: router/handler_edge_ctrl/apiSessionAdded.go [hunk 3]
```diff
@@ -93,61 +88,61 @@ func (h *apiSessionAddedHandler) HandleReceive(msg *channel2.Message, _ channel2
 				}
 
 				h.reqChan <- reqWithState
-			} else {
-				for _, session := range req.ApiSessions {
-					h.sm.AddApiSession(session)
+			} else if h.sm.IsSyncInProgress() {
+				reqWithState := &apiSessionAddedWithState{
```
Comment (andrewpmartinez) [inline router/handler_edge_ctrl/apiSessionAdded.go:92]: I assume the assumption is that if this sync fails, the ApiSession created during the sync will be present in the subsequent sync.

Comment (plorenz) [reply router/handler_edge_ctrl/apiSessionAdded.go:92]: Yes

**Review by andrewpmartinez — APPROVED** [review]

File: controller/server/controller.go (hunk 1 above)
Comment (andrewpmartinez) [reply controller/server/controller.go:266]: Too late, approved =(

Comment (plorenz) [reply controller/server/controller.go:266]: Nooooooooo!!

Comment (plorenz) [reply controller/server/controller.go:266]: We need some marker file that will fail the build with reminders of things to do before merge

File: controller/env/broker.go [hunk 4]
```diff
@@ -174,7 +182,10 @@ func (broker *Broker) apiSessionCertificateHandler(args ...interface{}) {
 	})
 
 	if err != nil {
-		pfxlog.Logger().WithError(err).Error("could not process API Session certificate event, failed to query for parent API Session")
+		// If it's not found, it's because it was deleted, which
```
Comment (andrewpmartinez) [inline controller/env/broker.go:185]: "which ..." ? Not critical but it leaves me on the edge of my seat.

**Review by andrewpmartinez — APPROVED** [review]

File: controller/env/broker.go (hunk 4 above)
Comment (plorenz) [reply controller/env/broker.go]: Tune in for the thrilling conclusion in next week's commit :smile: 

Fixed

**Pull request merged**